Issue #1654: ARRAY_EXTRACT for VARCHAR

hawkfish opened the pull request:
Implement and test plus a few bits of tidying up of `ARRAY_SLICE`.

Commit:
Issue #1654: ARRAY_EXTRACT for VARCHAR

Implement and test the function.

Commit:
Issue #1654: VARCHAR subscript syntax

Add tests for subscripting.

Commit:
Issue #1564: VARCHAR slice indexing

ARRAY_SLICE was not restricting its inputs
to 32 bits.

Commit:
Merge branch 'master' of https://github.com/duckdb/duckdb into duckdb-master

Commit:
Merge branch 'duckdb-master' into hawkfish-array-extract

Mytherin commented on test/sql/function/string/test_subscript.test:
```diff
@@ -0,0 +1,115 @@
+CREATE TABLE strings(s VARCHAR, off INTEGER);
+
+statement ok
+INSERT INTO strings VALUES ('hello', 0), ('world', 1), ('b', 0), (NULL, 1)
```
Could you add some tests involving unicode strings, and composite unicode characters (see e.g. `test/sql/function/string/test_complex_unicode.test`)

Mytherin commented on test/sql/function/string/test_subscript.test:
```diff
@@ -0,0 +1,115 @@
+
+# test direct subscript
+statement error
+SELECT '🦆ab'[0], 'abc'[1]
```
This should be fixed by #1661 :)

Mytherin commented on src/function/scalar/nested/list_extract.cpp:
```diff
@@ -124,23 +122,66 @@ static void ListExtractFunFun(DataChunk &args, ExpressionState &state, Vector &r
+		break;
+	case LogicalTypeId::VARCHAR:
+		// string extract returns a string, but can only accept 32 bit integers
+		bound_function.return_type = arguments[0]->return_type;
```
Rather than manually binding on the first type here, could you use the function binder by creating function overloads for LIST and VARCHAR, or does that not work with LIST arguments? Perhaps the binder needs to be adapted to support `LIST<ANY>` as an input parameter.

hawkfish replied:
I copied it from existing code, but I'll see if I can make this work.

hawkfish replied:
Yes it does. `CastRules::ImplicitCast` doesn't seem to understand `LIST` types.

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some minor comments:

hawkfish committed:
Merge pull request #52 from duckdb/master

Merge as of 2021-04-22

Commit:
Issue #1654: String constant subscripts

Update tests to reflect new subscripting syntax.
NULL::VARCHAR is still not supported.

Commit:
Issue #1654: Grapheme subscript tests

Add some tests with our favourite
nasty grapheme sequences.

Commit:
Issue #1654: ARRAY_EXTRACT binding

Fix the binding code so it will handle
LIST<ANY> inference correctly.

Mytherin commented:
Looks great, thanks!
